File: src/bots/exampleBot.py
import os
import sys
import random
import string
import time
import socketio
import signal

# ==== CONFIG ====
_NAMESPACE = '/bots'
SERVER_URL = os.getenv("BOT_SERVER_URL", "http://localhost:3000")
BOT_COUNT = int(os.getenv("BOT_COUNT", "3"))
BASE_NAME = os.getenv("BOT_NAME", "SamplePythonBot")

# ==== TYPES ====
class TurnState:
    def __init__(self, data):
        self.matchId = data.get("matchId")
        self.currentCard = data.get("currentCard")
        self.pot = data.get("pot", 0)
        self.you = data.get("you", {"chips": 0, "cards": []})

# ==== GLOBAL ====
bots = []

# ==== LOGIC ====
def choose_action(state: TurnState) -> str:
    if state.currentCard is None:
        return "take"
    if state.you.get("chips", 0) <= 0:
        return "take"

    cards = state.you.get("cards", [])
    min_card = min(cards) if cards else float("inf")
    potential_score = state.currentCard - state.pot
    if potential_score <= min_card:
        return "take"

    return "pass" if random.random() < 0.5 else "take"


def spawn_bot(name: str):
    sio = socketio.Client(
        reconnection=True,
        reconnection_delay=1,
        reconnection_delay_max=4,
    )

    bot = {"name": name, "socket": sio, "matchId": None}

    @sio.on("registered", namespace=_NAMESPACE)
    def on_registered(payload):
        stats = payload.get("stats")
        if stats:
            print(f"[{name}] stats {stats}")
        sio.emit("enqueue", namespace=_NAMESPACE)

    @sio.on("matchStarted", namespace=_NAMESPACE)
    def on_match_started(state):
        bot["matchId"] = state.get("matchId")
        players = ", ".join(p["name"] for p in state.get("players", []))
        print(f"[{name}] match started against {players}")

    @sio.on("turn", namespace=_NAMESPACE)
    def on_turn(data):
        if not data:
            return
        state = TurnState(data)
        bot["matchId"] = state.matchId
        decision = choose_action(state)
        print(f"[{name}] chooses {decision} (card {state.currentCard}, pot {state.pot}, chips {state.you.get('chips')}).")
        sio.emit("botAction", {"matchId": state.matchId, "action": decision}, namespace=_NAMESPACE)

    @sio.on("matchUpdate", namespace=_NAMESPACE)
    def on_match_update(state):
        if not state:
            return
        bot["matchId"] = state.get("matchId")

    @sio.on("matchEnded", namespace=_NAMESPACE)
    def on_match_ended(summary):
        if not summary:
            return
        standings = summary.get("standings", [])
        winners = summary.get("winners", [])
        placement = next((i + 1 for i, e in enumerate(standings) if e["name"] == name), None)
        score = next((e.get("totalScore", "n/a") for e in standings if e["name"] == name), "n/a")
        bot_id = next((e["botId"] for e in standings if e["name"] == name), "")
        win = bot_id in winners
        print(f"[{name}] match ended — place {placement}/{len(standings)} (score {score}){' ✅' if win else ''}")
        bot["matchId"] = None

    @sio.event(namespace=_NAMESPACE)
    def disconnect(reason):
        print(f"[{name}] disconnected: {reason}")

    @sio.event(namespace=_NAMESPACE)
    def connect_error(data):
        print(f"[{name}] connection error: {data}")

    sio.connect(f"{SERVER_URL}",  namespaces=["/bots"])
    sio.emit("registerBot", {"name": name}, namespace=_NAMESPACE)
    return bot


# ==== MAIN ====
def shutdown(signal_received=None, frame=None):
    print("\nShutting down bots…")
    for bot in bots:
        try:
            bot["socket"].disconnect()
        except Exception:
            pass
    time.sleep(0.2)
    sys.exit(0)


if __name__ == "__main__":
    signal.signal(signal.SIGINT, shutdown)

    for i in range(BOT_COUNT):
        suffix = ''.join(random.choices(string.ascii_lowercase + string.digits, k=3))
        name = f"{BASE_NAME}-{i+1}-{suffix}"
        bots.append(spawn_bot(name))

    while True:
        time.sleep(1)

File: src/bots/exampleBot.ts
/* eslint-disable no-console */
import { io, Socket } from "socket.io-client";

type TurnState = {
  matchId: string;
  currentCard: number | null;
  pot: number;
  you: {
    chips: number;
    cards: number[];
  };
};

const SERVER_URL = process.env.BOT_SERVER_URL || "http://localhost:3000";
const BOT_COUNT = Number.parseInt(process.env.BOT_COUNT || "3", 10);
const BASE_NAME = process.env.BOT_NAME || "SampleBot";
const BASE_SECRET =
  process.env.BOT_SECRET || `sample-secret-${Math.random().toString(36).slice(2, 10)}`;

interface SampleBot {
  name: string;
  socket: Socket;
  matchId: string | null;
  secret: string;
}

const bots: SampleBot[] = [];

for (let i = 0; i < BOT_COUNT; i += 1) {
  const name = `${BASE_NAME}-${i + 1}-${Math.random().toString(36).slice(2, 5)}`;
  const secret = BOT_COUNT > 1 ? `${BASE_SECRET}-${i + 1}` : BASE_SECRET;
  bots.push(spawnBot(name, secret));
}

function spawnBot(name: string, secret: string): SampleBot {
  const socket = io(`${SERVER_URL}/bots`, {
    transports: ["websocket", "polling"],
    reconnection: true,
    reconnectionDelay: 1000,
    reconnectionDelayMax: 4000,
  });

  const bot: SampleBot = {
    name,
    socket,
    matchId: null,
    secret,
  };

  socket.on("connect", () => {
    console.log(`[${name}] connected, registering…`);
    socket.emit("registerBot", { name, secret }, (ack?: { ok?: boolean; error?: string; rating?: number }) => {
      if (!ack?.ok) {
        console.error(`[${name}] registration failed: ${ack?.error || "unknown error"}`);
        return;
      }
      console.log(`[${name}] ready (rating ${ack.rating}).`);
      socket.emit("enqueue");
    });
  });

  socket.on("registered", (payload: { stats?: unknown }) => {
    if (payload?.stats) {
      console.log(`[${name}] stats`, payload.stats);
    }
  });

  socket.on("matchStarted", (state: { matchId?: string; players: Array<{ name: string }> }) => {
    bot.matchId = state.matchId ?? null;
    console.log(
      `[${name}] match started against ${state.players.map((player) => player.name).join(", ")}`,
    );
  });

  socket.on("turn", (state: TurnState) => {
    if (!state) {
      return;
    }
    bot.matchId = state.matchId;
    const decision = chooseAction(state);
    console.log(
      `[${name}] chooses ${decision} (card ${state.currentCard}, pot ${state.pot}, chips ${state.you.chips}).`,
    );
    socket.emit("botAction", { matchId: state.matchId, action: decision });
  });

  socket.on("matchUpdate", (state: { matchId?: string }) => {
    if (!state) {
      return;
    }
    bot.matchId = state.matchId ?? null;
  });

  socket.on("matchEnded", (summary: {
    standings: Array<{ name: string; totalScore?: number; botId: string }>;
    winners: string[];
  }) => {
    if (!summary) {
      return;
    }
    const placement =
      summary.standings.findIndex((entry) => entry.name === name) + 1;
    const score =
      summary.standings.find((entry) => entry.name === name)?.totalScore ?? "n/a";
    const win = summary.winners.includes(
      summary.standings.find((entry) => entry.name === name)?.botId ?? "",
    );
    console.log(
      `[${name}] match ended — place ${placement}/${summary.standings.length} (score ${score})${
        win ? " ✅" : ""
      }`,
    );
    bot.matchId = null;
  });

  socket.on("disconnect", (reason: string) => {
    console.log(`[${name}] disconnected: ${reason}`);
  });

  socket.on("error", (error: unknown) => {
    console.error(`[${name}] socket error:`, error);
  });

  return bot;
}

function chooseAction(state: TurnState): "pass" | "take" {
  if (!state || state.currentCard == null) {
    return "take";
  }

  if (state.you.chips <= 0) {
    return "take";
  }

  const minCard = Math.min(...(state.you.cards.length ? state.you.cards : [Infinity]));
  const potentialScore = state.currentCard - state.pot;
  if (potentialScore <= minCard) {
    return "take";
  }

  return Math.random() < 0.5 ? "pass" : "take";
}

process.on("SIGINT", () => {
  console.log("\nShutting down bots…");
  bots.forEach((bot) => {
    try {
      bot.socket.disconnect();
    } catch {
      // ignore
    }
  });
  setTimeout(() => process.exit(0), 200);
});
